src/main/java/core/basesyntax/service/impl/ProductCounterServiceImpl.java

```
@@ -0,0 +1,58 @@
+import java.util.List;
+
+public class ProductCounterServiceImpl implements ProductCounterService {
+    private final ActionService actionService;
+
+    public ProductCounterServiceImpl(ActionService actionService) {
+        if (actionService == null) {
+            throw new ActionServiceIsNullException("ActionService can't be null");
+        }
+        this.actionService = actionService;
+    }
+
+    @Override
+    public void countTheProducts(List<FruitTransaction> fruitTransactions, Storage storage) {
+        if (fruitTransactions == null) {
+            throw new ReportIsNullException("Report can't be null");
+        }
+        if (storage == null) {
+            throw new StorageIsNullException("Storage can't be null");
+        }
+        if (fruitTransactions.size() <= 1) {
+            return;
+        }
+        for (FruitTransaction fruitTransaction : fruitTransactions) {
+            String productName = fruitTransaction.getProductName();
+            int quantity = fruitTransaction.getQuantity();
+            String actionCode = fruitTransaction.getActionCode();
+            if (actionCode == null || productName == null) {
+                throw new IncorrectFormatOfDataException("ActionCode"
+                        + " and ProductName can't be null");
+            }
+            if (actionCode.isEmpty() ||  productName.isEmpty()) {
+                throw new IncorrectFormatOfDataException(
+                        "Fruit transaction: action/product must not be blank -> " +
+                                fruitTransaction
+                );
+            }
+            if (quantity < 0) {
+                throw new QuantityLessThanNullException("Quantity must be "
+                        + "greater than or equal to 0: " + quantity);
+            }
+            ActionTypeService action = actionService.getStrategy(actionCode);
+            action.applyTheQuantity(storage, productName, quantity);
```

This violates checklist items about validating and normalizing operation codes: "ActionService and Operation: `Operation.fromCode` should null-check and trim the input, throwing `InvalidDataException` on null/unknown codes." and "Validate operation codes; throw informative exception for unknown or null operation."

In this loop you check `actionCode` with `isEmpty()` and `isEmpty()`/`isEmpty` tests but do not `trim()` the values nor enforce that `actionCode` has been normalized/mapped to an `Operation` enum before calling `actionService.getStrategy(actionCode)` (lines 40–55). Relying on raw strings here risks mismatches due to whitespace/casing and defers responsibility for validation/mapping from the parser/model to this service. Move trimming/validation/mapping to the parser or `Operation.fromCode()` and ensure `actionService.getStrategy(...)` receives a validated `Operation` or trimmed code; also ensure that unknown operations throw informative exceptions rather than returning null or causing obscure errors.